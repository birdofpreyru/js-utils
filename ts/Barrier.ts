type Resolver<T> = (value: T | PromiseLike<T>) => void;
type Rejecter = (reason?: any) => void;
type Executor<T> = (resolve: Resolver<T>, reject: Rejecter) => void;

enum STATE {
  PENDING = 'PENDING',
  REJECTED = 'REJECTED',
  RESOLVED = 'RESOLVED',
}

/**
 * Barrier is just a Promise which has resolve and reject exposed as instance
 * methods.
 *
 * Docs: https://dr.pogodin.studio/docs/react-utils/docs/api/classes/Barrier
 */
export class Barrier<T, TR = T> extends Promise<TR> {
  private resolver: Resolver<T> | Resolver<TR>;

  private rejecter: Rejecter;

  private state = STATE.PENDING;

  constructor(executor?: Executor<TR>) {
    let resolveRef: Resolver<TR>;
    let rejectRef: Rejecter;

    super((resolve, reject) => {
      resolveRef = (value: TR | PromiseLike<TR>) => {
        resolve(value);
        this.state = STATE.RESOLVED;
      };
      rejectRef = (reason?: any) => {
        reject(reason);
        this.state = STATE.REJECTED;
      };
      if (executor) executor(resolveRef, rejectRef);
    });

    this.resolver = resolveRef!;
    this.rejecter = rejectRef!;
  }

  get resolve() { return <Resolver<T>> this.resolver; }

  get reject() { return this.rejecter; }

  get resolved() { return this.state === STATE.RESOLVED; }

  get rejected() { return this.state === STATE.REJECTED; }

  get settled() { return this.state !== STATE.PENDING; }

  catch<TR1>(
    onRejected?: ((reason: any) => TR1 | PromiseLike<TR1>) | null,
  ): Barrier<T, TR1> {
    return <Barrier<T, TR1>> super.catch(onRejected);
  }

  finally(onFinally?: (() => void) | null): Barrier<TR> {
    return <Barrier<TR>> super.finally(onFinally);
  }

  then<TR1, TR2>(
    onFulfilled?: ((value: TR) => TR1 | PromiseLike<TR1>) | null,
    onRejected?: ((reason: any) => TR2 | PromiseLike<TR2>) | null,
  ): Barrier<T, TR1 | TR2> {
    const res = <Barrier<T, TR1 | TR2>> super.then(onFulfilled, onRejected);
    res.resolver = this.resolve;
    res.rejecter = this.reject;
    return res;
  }
}

/**
 * Creates a new Barrier.
 * @returns {Barrier}
 */
export function newBarrier<T>(executor?: Executor<T>) {
  return new Barrier<T>(executor);
}
