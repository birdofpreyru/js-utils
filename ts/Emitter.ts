type Listener = (...args: any[]) => void;

/**
 * Simple listeneable data Emitter.
 */
export default class Emitter {
  private p_listeners: Listener[] = [];

  /**
   * Returns "true" if any listener is connected; "false" otherwise.
   * @return {boolean}
   */
  get hasListeners(): boolean {
    return !!this.p_listeners.length;
  }

  get listeners(): ReadonlyArray<Listener> { return this.p_listeners; }

  /**
   * Adds `listener` if it is not already connected.
   * @param {function} listener
   * @return {function} Unsubscribe function.
   */
  addListener(listener: Listener): () => void {
    if (!this.p_listeners.includes(listener)) {
      this.p_listeners.push(listener);
    }
    return () => this.removeListener(listener);
  }

  /**
   * Calls every connected listener with the given arguments.
   * @param  {...any} args
   */
  emit(...args: any[]) {
    const { p_listeners: listeners } = this;
    for (let i = 0; i < listeners.length; ++i) {
      listeners[i](...args);
    }
  }

  /**
   * Removes specified `listener`, if connected.
   * @param {function} listener
   */
  removeListener(listener: Listener) {
    const idx = this.p_listeners.indexOf(listener);
    if (idx >= 0) this.p_listeners.splice(idx, 1);
  }
}
