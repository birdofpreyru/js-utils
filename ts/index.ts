export { Barrier, newBarrier } from './Barrier';
