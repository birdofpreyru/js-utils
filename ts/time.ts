import Barrier, { type Executor } from './Barrier';

export const SEC_MS = 1000;
export const MIN_MS = 60 * SEC_MS;
export const HOUR_MS = 60 * MIN_MS;
export const DAY_MS = 24 * HOUR_MS;
export const YEAR_MS = 365 * DAY_MS;

class Timer<T> extends Barrier<void, T> {
  private p_abort: () => void;

  get abort(): () => void { return this.p_abort; }

  constructor(executor?: Executor<T>, timeout: number = 0) {
    super(executor);
    if (timeout > 0) {
      const id = setTimeout(super.resolve.bind(this), timeout);
      this.p_abort = () => clearTimeout(id);
    } else {
      this.p_abort = () => {};
      super.resolve();
    }
  }

  then<TR1, TR2>(
    onFulfilled?: ((value: T) => TR1 | PromiseLike<TR1>) | null,
    onRejected?: ((reason: any) => TR2 | PromiseLike<TR2>) | null,
  ): Timer<TR1 | TR2> {
    const res = <Timer<TR1 | TR2>> super.then(onFulfilled, onRejected);
    res.p_abort = this.p_abort;
    return res;
  }
}

/**
 * Creates a Promise, which resolves after the given timeout.
 * @param {number} timeout Timeout [ms].
 * @return {Barrier} Resolves after the timeout. It has additional
 *  .abort() method attached, which cancels the pending timer resolution
 *  (without resolving or rejecting the barrier).
 */
export async function timer(timeout: number) {
  return new Timer<void>(undefined, timeout);
}
