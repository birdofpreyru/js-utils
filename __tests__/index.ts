import * as lib from '../ts';

it('exports expected definitions', () => {
  expect(lib).toMatchSnapshot();
});
