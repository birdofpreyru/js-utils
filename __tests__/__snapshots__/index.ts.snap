// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports expected definitions 1`] = `
{
  "Barrier": [Function],
  "DAY_MS": 86400000,
  "Emitter": [Function],
  "HOUR_MS": 3600000,
  "MIN_MS": 60000,
  "SEC_MS": 1000,
  "Semaphore": [Function],
  "Timer": [Function],
  "YEAR_MS": 31536000000,
  "timer": [Function],
}
`;
